HW/iris.py

```diff
@@ -1,75 +1,142 @@
+def iris_plots():
+    # scatter plot of sepal_width vs petal length, color shows the sepal length
+    iris_data = import_iris_data()
+    iris_scatter = px.scatter(
+        iris_data, x="sepal_width", y="petal_length", color="sepal_length"
+    )
+    iris_scatter.show()
+    # violin plot of sepal_width
+    iris_violin = px.violin(iris_data, y="sepal_width")
+    iris_violin.show()
+    # bar plots for sepal_width and sepal_length
+    iris_bar = px.bar(iris_data, x="sepal_width", y="sepal_length")
+    iris_bar.show()
+    # density heatmap of sepal_length vs sepal_width
+    iris_heatmap = px.density_heatmap(iris_data, x="sepal_length", y="sepal_width")
+    iris_heatmap.show()
+    # sepal_length box plot
+    iris_box_plot = px.box(iris_data, y="sepal_length")
+    iris_box_plot.show()
+
+
+def test_train():
+    # RF model
+    # defining features dataset without labels
+    iris_data = import_iris_data()
+    features = iris_data.iloc[:, 0:3].values
+    # Labels are the values we want to predict
+    labels = iris_data.iloc[:, 4].values
+    # Split the data into training and testing sets
+    train_features, test_features, train_labels, test_labels = train_test_split(
+        features, labels, test_size=0.2, random_state=0
+    )
+    return train_features, test_features, train_labels, test_labels
+
+
+def predict_model():
+    train_features, test_features, train_labels, test_labels = test_train()
+    rf_model_pipeline = Pipeline(
+        steps=[
+            ("scaling", StandardScaler()),
+            ("random_forest", RandomForestClassifier(n_estimators=20, random_state=0)),
+        ]
+    )
+    rf_model_pipeline.fit(train_features, train_labels)
+    rf_model_pipeline_predictions = rf_model_pipeline.predict(test_features)
+    print("-----RF Model Prediction------")
+    print(
+        "confustion matrix is : ",
+        confusion_matrix(test_labels, rf_model_pipeline_predictions),
+    )
+    print(
+        "calssification report is :",
+        classification_report(test_labels, rf_model_pipeline_predictions),
+    )
+    print(
+        "Accuracy score is:", accuracy_score(test_labels, rf_model_pipeline_predictions)
+    )
+
+    KNN_model_pipeline = Pipeline(
+        steps=[
+            ("scaling", StandardScaler()),
+            ("KNN", KNeighborsClassifier(n_neighbors=8)),
+        ]
+    )
+    KNN_model_pipeline.fit(train_features, train_labels)
+    KNN_model_pipeline_predictions = rf_model_pipeline.predict(test_features)
+    print("-----KNN Model Prediction------")
+    print(
+        "confustion matrix is : ",
+        confusion_matrix(test_labels, KNN_model_pipeline_predictions),
+    )
+    print(
+        "calssification report is :",
+        classification_report(test_labels, KNN_model_pipeline_predictions),
+    )
+    print(
+        "Accuracy score is:",
+        accuracy_score(test_labels, KNN_model_pipeline_predictions),
+    )
+
+
+def main():
+    import_iris_data()
```

Comment:
So you're calling some functions down here but not saving the returned output. Then you call them again in the next function. It works to get the plots generated for me but it's definitely running the function twice.